[DAL] Spec of common types (#222)
oneapi-src/oneAPI-spec: source/elements/oneDAL/source/common_interface/common_type_definitions.rst

=======================
Common type definitions
=======================

This section describes common types used in |dal_short_name|.

---------------------
Programming interface
---------------------

All types and functions in this section shall be declared in the
``oneapi::dal`` namespace and be available via inclusion of the
``oneapi/dal/common.hpp`` header file.

.. _scalar_types:

Scalar types
------------

|dal_short_name| relies on standard C++ scalar types defined in ``<cstdint>``.
This file shall be included in ``oneapi/dal/common.hpp`` and all

> |dal_short_name| types shall use this scalar types.

Reviewer: ```suggestion
|dal_short_name| types shall use these scalar types.
```

For variables indicating size, amount, count of entities, ``std::int64_t`` type
shall be used.

It is recommended to use standard C++ types for applications as well.

Enum classes
------------

> It is up to |dal_short_name| implementer when to set particular base type of
> ``enum`` or ``enum class`` type implementing |dal_short_name| concept unless it
> is explicitly specified.

Reviewer: ```suggestion
Unless it is explicitly specified, it is up to |dal_short_name| implementer to decide
when to set a particular base type of ``enum`` or ``enum class`` type implementing |dal_short_name| concept.
```

Data type
---------

The implementation of :capterm:`data type` concept. It shall enumerate all the
data types supported by |dal_short_name| to perform computations. The

> ``data_type`` class shall contain all the base :txtref:`scalar_types` and can
> extend it. Base scalar types include all which names follow the pattern
> ``std::int_XX_t`` or ``std::uint_XX_t``, where XX is one of 8, 16, 32, or 64.

Reviewer: ```suggestion
``data_type`` class shall contain all the base :txtref:`scalar_types` and can also
extend them. Base scalar types include the types whose names follow the pattern
``std::int_XX_t`` or ``std::uint_XX_t``, where XX is 8, 16, 32, or 64.
```

.. onedal_enumclass:: oneapi::dal::data_type

Range
-----

> The range ``[start_index, end_index)`` in the array or another container that
> supports value indexing.

Reviewer: ```suggestion
A range ``[start_index, end_index)`` in an array or any other container that
supports value indexing.
```

Scalar types
------------

|dal_short_name| relies on the use of integral types defined in ``<cstdint>``.
This file shall be included in ``oneapi/dal/common.hpp`` and all
|dal_short_name| types shall use these data types.

> The interfaces of the library shall use ``std::int64_t`` data type to represent
> problem's sizes such as the number of rows and columns in the table.

Reviewer: ```suggestion
The interfaces of the library shall use ``std::int64_t`` data type to represent
dimensionality (for example, the number of rows and columns in the table).
```

or without "for example"

It is recommended to use standard C++ types for applications as well.

Enum classes
------------

> Which base type to use when defining ``enum`` or ``enum class`` representing

Reviewer: ```suggestion
Which base type to use when defining ``enum`` or ``enum class`` representing a
```
